src/create-subscription-orders.js

```diff
@@ -0,0 +1,243 @@
+import fetch from 'isomorphic-fetch'
+import pMap from 'p-map'
+import parser from 'cron-parser'
+import { getSubscriptionConfig } from './config.js'
+import { updateOrderWithRetry } from './utils/utils.js'
+import {
+  ACTIVE_STATE,
+  ERROR_STATE,
+  REMINDER_SENT_STATE,
+  SEND_REMINDER_STATE,
+} from './states-constants.js'
+
+let apiRoot
+let ctpClient
+let logger
+let stats
+
+async function createSubscriptionOrders({
+  apiRoot: _apiRoot,
+  ctpClient: _ctpClient,
+  logger: _logger,
+  stateKeyToIdMap,
+}) {
+  stats = {
+    processedTemplateOrders: 0,
+    subscriptionOrderAlreadyCreated: 0,
+    successUrlCalls: 0,
+    recoverableErrorUrlCalls: 0,
+    unrecoverableErrorUrlCalls: 0,
+  }
+  apiRoot = _apiRoot
+  ctpClient = _ctpClient
+  logger = _logger
+  const stateIds = [
+    stateKeyToIdMap.get(ACTIVE_STATE),
+    stateKeyToIdMap.get(SEND_REMINDER_STATE),
+    stateKeyToIdMap.get(REMINDER_SENT_STATE),
+  ]
+  const orderQuery =
+    await _buildQueryForTemplateOrdersThatNeedSubscriptionOrders(stateIds)
+
+  for await (const templateOrders of ctpClient.fetchPagesGraphQl(orderQuery))
+    await pMap(templateOrders, _processTemplateOrder, { concurrency: 3 })
+
+  return stats
+}
+
+async function _processTemplateOrder({
+  id,
+  version,
+  orderNumber,
+  custom: { customFieldsRaw },
+  state: { key: stateKey },
+}) {
+  const config = getSubscriptionConfig()
+  const headers = _buildHeaders(config)
+  const subscriptionOrderCreationUrl = config.subscriptionOrderCreationUrl
+
+  const deliveryDate = customFieldsRaw.find(
+    (attr) => attr.name === 'nextDeliveryDate'
+  ).value
+  const subscriptionOrderExists =
+    await _fetchOrderBySubscriptionTemplateOrderRefAndDeliveryDate(
+      id,
+      deliveryDate
+    )
+
+  if (subscriptionOrderExists) {
+    logger.info(
+      `Template order ${orderNumber}: ` +
+        `Subscription order was already created for date ${deliveryDate}. Will update the template order only.`
+    )
+    stats.subscriptionOrderAlreadyCreated++
+    await _updateTemplateOrder(id, version, customFieldsRaw, stateKey)
+  } else {
+    const response = await fetch(subscriptionOrderCreationUrl, {
```

Comment:
I think if node-fetch fails it will throw an error, can you check that 🤔 ?

Reply:
Normally it should not fail if it's not a code error (e.g. missing URL). If it's 4xx or 5xx errors, it is not an error: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
```
The Promise returned from fetch() won't reject on HTTP error status even if the response is an HTTP 404 or 500. Instead, as soon as the server responds with headers, the Promise will resolve normally (with the [ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) property of the response set to false if the response isn't in the range 200–299), and it will only reject on network failure or if anything prevented the request from completing.
```

Reply:
Thanks for checking it!